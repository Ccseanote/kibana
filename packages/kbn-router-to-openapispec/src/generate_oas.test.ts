/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { generateOpenApiDocument } from './generate_oas';
import { CoreVersionedRouter, Router } from '@kbn/core-http-router-server-internal';
import { schema } from '@kbn/config-schema';

const createRouter = (args: { routes: ReturnType<Router['getRoutes']> }): Router => {
  const router: Router = Object.create(Router);
  return {
    router,
    getRoutes: () => args.routes,
  } as unknown as Router;
};
const createVersionedRouter = (args: {
  routes: ReturnType<CoreVersionedRouter['getRoutes']>;
}): CoreVersionedRouter => {
  return {
    getRoutes: () => args.routes,
  } as unknown as CoreVersionedRouter;
};

describe('generateOpenApiDocument', () => {
  describe('happy path', () => {
    it('generates the expected OpenAPI document', () => {
      expect(
        generateOpenApiDocument(
          {
            routers: [
              createRouter({
                routes: [
                  {
                    isVersioned: false,
                    path: '/foo',
                    method: 'get',
                    validationSchemas: {
                      body: schema.object({ foo: schema.string() }),
                    },
                    options: {
                      tags: ['foo'],
                      responses: {
                        200: {
                          body: schema.object({ fooResponse: schema.string() }),
                        },
                      },
                    },
                    handler: jest.fn(),
                  },
                ],
              }),
            ],
            versionedRouters: [
              createVersionedRouter({
                routes: [
                  {
                    method: 'get',
                    path: '/bar',
                    options: {
                      access: 'public',
                    },
                    handlers: [
                      {
                        fn: jest.fn(),
                        options: {
                          validate: {
                            request: { body: schema.object({ foo: schema.string() }) },
                            response: {
                              [200]: { body: schema.object({ fooResponse: schema.string() }) },
                            },
                          },
                          version: 'oas-test-version-1',
                        },
                      },
                      {
                        fn: jest.fn(),
                        options: {
                          validate: {
                            request: { body: schema.object({ foo: schema.string() }) },
                            response: {
                              [200]: { body: schema.object({ fooResponse: schema.string() }) },
                            },
                          },
                          version: 'oas-test-version-2',
                        },
                      },
                    ],
                  },
                ],
              }),
            ],
          },
          {
            title: 'test',
            baseUrl: 'https://test.oas',
            version: '99.99.99',
          }
        )
      ).toMatchSnapshot();
    });
  });
});
