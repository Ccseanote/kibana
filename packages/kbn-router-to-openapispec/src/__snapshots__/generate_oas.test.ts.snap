// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateOpenApiDocument happy path generates the expected OpenAPI document 1`] = `
Object {
  "externalDocs": undefined,
  "info": Object {
    "description": undefined,
    "title": "test",
    "version": "99.99.99",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/bar": Object {
      "get": Object {
        "operationId": "/bar#0",
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json; Elastic-Api-Version=oas-test-version-1": Object {
              "schema": Object {
                "additionalProperties": false,
                "properties": Object {
                  "foo": Object {
                    "type": "string",
                  },
                },
                "required": Array [
                  "foo",
                ],
                "type": "object",
              },
            },
            "application/json; Elastic-Api-Version=oas-test-version-2": Object {
              "schema": Object {
                "additionalProperties": false,
                "properties": Object {
                  "foo": Object {
                    "type": "string",
                  },
                },
                "required": Array [
                  "foo",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json; Elastic-Api-Version=oas-test-version-1": Object {
                "schema": Object {
                  "additionalProperties": false,
                  "properties": Object {
                    "fooResponse": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "fooResponse",
                  ],
                  "type": "object",
                },
              },
              "application/json; Elastic-Api-Version=oas-test-version-2": Object {
                "schema": Object {
                  "additionalProperties": false,
                  "properties": Object {
                    "fooResponse": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "fooResponse",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "No description",
          },
        },
      },
    },
    "/foo": Object {
      "get": Object {
        "operationId": "/foo#0",
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "additionalProperties": false,
                "properties": Object {
                  "foo": Object {
                    "type": "string",
                  },
                },
                "required": Array [
                  "foo",
                ],
                "type": "object",
              },
            },
          },
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "additionalProperties": false,
                  "properties": Object {
                    "fooResponse": Object {
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "fooResponse",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "No description",
          },
        },
      },
    },
  },
  "security": Array [
    Object {
      "basicAuth": Array [],
    },
    Object {
      "apiKeyAuth": Array [],
    },
  ],
  "servers": Array [
    Object {
      "url": "https://test.oas",
    },
  ],
  "tags": undefined,
}
`;
